### PR TITLE
Add EntityBuilderClone::add_bundle and tests

### PR DESCRIPTION
**Pull request by ten3roberts**
This fixes issue #211 without any apparent breakages in the public API, allowing for an immediate patch release.

The commit also adds appropriate tests for the cloneable entity builder which were previously missing.

### CODE REVIEW

File: src/entity_builder.rs [hunk 1]
```diff
@@ -374,15 +387,24 @@ impl<M> Default for Common<M> {
...
-        Self(|src, f| unsafe {
-            let mut tmp = (*src.cast::<T>()).clone();
-            f((&mut tmp as *mut T).cast(), TypeInfo::of::<T>());
-            core::mem::forget(tmp);
-        })
+    pub(crate) fn new<T: Component + Clone>(type_info: TypeInfo) -> Self {
+        Self {
+            type_info,
+            func: |src, ty, f| unsafe {
+                std::dbg!(ty);
```
Comment (Ralith) [inline src/entity_builder.rs:402]: Stray `dbg`

File: src/world.rs [hunk 2]
```diff
@@ -159,6 +159,7 @@ impl World {
             components.put(|ptr, ty| {
                 archetype.put_dynamic(ptr, ty.id(), ty.layout().size(), index);
             });
+
```
Comment (Ralith) [inline src/world.rs:162]: Stray edit

File: tests/tests.rs [hunk 3]
```diff
@@ -200,6 +200,31 @@ fn build_entity() {
     assert_eq!(*world.get::<i32>(f).unwrap(), 789);
 }
 
+#[test]
+fn build_entity_clone() {
+    let mut world = World::new();
+    let mut entity = EntityBuilderClone::new();
+    entity.add("abc");
+    entity.add(123);
+    let e = world.spawn(&entity.build());
```
Comment (Ralith) [inline tests/tests.rs:209]: What's the test coverage benefit of constructing two different builders?

File: tests/tests.rs [hunk 4]
```diff
@@ -200,6 +200,31 @@ fn build_entity() {
...
+    let mut entity = EntityBuilderClone::new();
+    entity.add("def");
+    entity.add([0u8; 1024]);
+    entity.add(456);
+    entity.add(789);
+    entity.add_bundle(("yup", 67_usize));
+    entity.add_bundle((5.0_f32, String::from("Foo")));
+    entity.add_bundle((7.0_f32, String::from("Bar"), 42_usize));
+    let f = world.spawn(&entity.build());
```
Comment (Ralith) [inline tests/tests.rs:218]: Let's spawn multiple times so that any inadvertent memcpying is more likely to trigger a crash.

File: src/entity_builder.rs [hunk 5]
```diff
@@ -374,15 +387,24 @@ impl<M> Default for Common<M> {
 }
 
 #[derive(Clone)]
-struct Cloneable(unsafe fn(*const u8, &mut dyn FnMut(*mut u8, TypeInfo)));
+/// Struct providing dynamic clone via callback
+pub struct Cloneable {
+    func: unsafe fn(*const u8, TypeInfo, &mut dyn FnMut(*mut u8, TypeInfo)),
+    // Store type info along with cloneable
+    type_info: TypeInfo,
```
Comment (Ralith) [inline src/entity_builder.rs:394]: Why the change to store this rather than constructing it on-demand?

**Review by Ralith — CHANGES_REQUESTED** [review]
Thanks! I agree this is a very natural extension.

File: src/entity_builder.rs (hunk 5 above)
Comment (ten3roberts) [reply src/entity_builder.rs:394]: You are right. If typeinfo is cheap to construct it is not necessary to store it.

Comment (ten3roberts) [reply src/entity_builder.rs:394]: It was left over from storing the type info in an earlier attempt. Since it is known with bundle::put I thought to store it rather than compute again.

**Review by ten3roberts — COMMENTED** [review]
Great at spotting these.

File: tests/tests.rs (hunk 3 above)
Comment (ten3roberts) [reply tests/tests.rs:209]: This follows the style of `built_entity` test. It may increase the coverage as one with many modifications and one with few modifications of already existing components. Will add testing for spawning multiple entities from the same builder

File: src/entity_builder.rs (hunk 5 above)
Comment (ten3roberts) [reply src/entity_builder.rs:394]: The function is from my understanding called for each build

Comment (Ralith) [reply src/entity_builder.rs]: `TypeInfo::of` will almost certainly be compiled out into a constant, so storing `TypeInfo`s will probably *add* overhead, in addition to complicating the code. As usual, beware premature optimizations.

File: src/bundle.rs [hunk 6]
```diff
@@ -57,6 +58,17 @@ pub unsafe trait Bundle: DynamicBundle {
         Self: Sized;
 }
 
+/// A dynamically typed collection of components that is also cloneable
+pub unsafe trait DynamicBundleClone: DynamicBundle + Clone {
```
Comment (Ralith) [inline src/bundle.rs:62]: I think this is slightly more strict than necessary; we don't actually need to be able to clone the bundle itself, we just need to know how to clone its contents.
```suggestion
/// A dynamically typed collection of cloneable components
pub unsafe trait DynamicBundleClone: DynamicBundle {
```

**Review by Ralith — CHANGES_REQUESTED** [review]

File: src/entity_builder.rs [hunk 7]
```diff
@@ -173,6 +174,18 @@ impl EntityBuilderClone {
         self
     }
 
+    /// Add all components in `bundle` to the entity.
+    ///
+    /// If the bundle contains any component which matches the type of a component
+    /// already in the `EntityBuilder`, the newly added component from the bundle
+    /// will replace the old component and the old component will be dropped.
+    pub fn add_bundle<T: Component + DynamicBundleClone>(&mut self, bundle: T) -> &mut Self {
```
Comment (Ralith) [inline src/entity_builder.rs:182]: `Component` is overly strict here, and let's use `impl Trait` for consistency with `EntityBuilder`.
```suggestion
    pub fn add_bundle(&mut self, bundle: impl DynamicBundleClone) -> &mut Self {
```

**Review by Ralith — CHANGES_REQUESTED** [review]

File: src/bundle.rs (hunk 6 above)
Comment (ten3roberts) [reply src/bundle.rs:62]: That is correct. Can't think of an implementation where those aren't transitive properties unless a custom implementation.

**Comment by ten3roberts** [comment]
All tests pass after introducing the changes

**Review by Ralith — APPROVED** [review]
Thanks!
